### ISSUE TITLE
Feature: Ability to cache/store music offline

### ISSUE THREAD

**Issue by neomodern** [tomahawk-player/tomahawk#569]
Useful when having no internet access, poor speed or bandwidth.

**Comment by changemachine** [comment]
I was thinking the same thing.  Why waste bandwidth+CPU+memory on downloading something we already downloaded, and are likely to download again (and again, and again...).  Songs could simply be stored in a rolling cache.

**Comment by muesli** [comment]
We'd love to support that, but it's obviously a legal minefield. Some plugins like Spotify do their own internal caching though.

**Comment by X-Ryl669** [comment]
Could it be possible to add something similar to the "Collection" interface that's shared between C++ and JS, so that resolvers could use that to cache the songs' data ? 
By default, I don't see any legal issue to have my own Ampache/Owncloud songs cached on my own phone, yet, downloading songs numerous time is a PITA with a sparse cellular network.

**Comment by changemachine** [comment]
Right.  Everything is cached somewhere, it's just not something TH is controlling or accessing.  It might be, as muesli said, a "legal minefield", but not for all sources.

**Comment by MWrathDev** [comment]
Might be worth researching which services would allow such action? I'll start, pretty sure youtube is out - in fact even the current youtube functionality of TH might be in violation:

https://www.youtube.com/t/terms

## 4. General Use of the Service—Permissions and Restrictions

> C. You agree not to access Content through any technology or means other than the video playback pages of the Service itself, the Embeddable Player, or other explicitly authorized means YouTube may designate.

## 6. Your Content and Conduct

> B. Content is provided to you AS IS. You may access Content for your information and personal use solely as intended through the provided functionality of the Service and as permitted under these Terms of Service. **You shall not download any Content unless you see a “download” or similar link displayed by YouTube on the Service for that Content. You shall not copy, reproduce, make available online or electronically transmit, publish, adapt, distribute, transmit, broadcast, display, sell, license, or otherwise exploit any Content for any other purposes without the prior written consent of YouTube or the respective licensors of the Content.**

**Comment by MWrathDev** [comment]
Jamendo content may be cached:

https://www.jamendo.com/legal/terms-of-use

## CLAUSE 1. DEFINITIONS

> WORK(S): shall mean the musical works made available to Users of JAMENDO by Member Artists.

## CLAUSE 3. USE OF SERVICES

### 3.2 - Technical Information

> The Works may be downloaded in MP3 format and may be offered in different formats from time to time depending on the evolution of the Services. Third-party software may be required for the downloading of the Works.

**Comment by MWrathDev** [comment]
Last.fm audio may not be cached:

http://www.last.fm/legal/terms

## 3. Intellectual Property Ownership; Licence

> The copying, reproduction, re-arrangement, sale, leasing, renting, lending, distribution, redistribution, modification or adaptation, downloading, sideloading, exchanging, creating of derivative works, uploading, posting, transmitting, communication to the public or publication by you, directly or indirectly, of the Content, including the removal or alteration of advertising, except pursuant to the express limited grant of rights hereunder, is strictly prohibited.

> do not copy or seek to copy or “rip” any audio and/or audiovisual content from the Website or any part of the Service (including, without limitation, the Last.fm widget);

**Comment by MWrathDev** [comment]
Bandcamp audio may be cached:

https://bandcamp.com/terms_of_use

## Intellectual Property Rights – Artists.

> (ii) allow users of the Service to receive public performances and public displays of the Artist’s Music and Artworks and to **reproduce the Artist’s Music and Artworks on any and all devices owned or controlled by the user for non-commercial purposes** and receive performances and displays of same;

**Comment by MWrathDev** [comment]
Soundcloud audio may be cached given certain circumstances:

https://soundcloud.com/terms-of-use

## Your use of the Platform

> Subject to your strict compliance with these Terms of Use and our Community Guidelines at any and all times during your use of the Platform, SoundCloud grants you a limited, personal, non-exclusive, revocable, non-assignable and non-transferable right and license to use the Platform in order to view Content uploaded and posted to the Website, **to listen to audio Content streamed from the Platform or offline** and to share and download audio Content using the features of the Platform where the appropriate functionality has been enabled by the user who uploaded the relevant Content (the “Uploader”), and subject to the territorial availability of that feature and audio Content.

> (i) You must not copy, rip or capture, or attempt to copy, rip or capture, any audio Content from the Platform or any part of the Platform, other than by means of download or store for offline listening in circumstances where the relevant Uploader has elected to permit downloads or offline listening of the relevant item of Content.

**Comment by MWrathDev** [comment]
That's most of the free streaming services, (couldnt access Official.fm), ill let someone else do the subscription and network/cloud ones.

The only way around such things may be non-profit / fair use, but someone else can also look into that, lookin forward to it though, IMHO this could be the music player to end all music players.

**Comment by muesli** [comment]
Nice work @MWrathDev !

**Comment by CallMeRev** [comment]
dam i just istalled this app, too bad you cant use offline that the reason i wanted it , i have spotify premium , soundcloud and bandcamp songs as well but its anoying to cant play everything offline in the same app...
